sql/codegen/intermediate_representation.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2019 The SQLFlow Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codegen
+
+import "github.com/sql-machine-learning/sqlflow/sql/columns"
+
+// FieldType indicates the field type of a table column
+type FieldType int
+
+const (
+	// Int indicates the corresponding table column is an integer
+	Int FieldType = iota
+	// Float indicates the corresponding table column is a float
+	Float
+	// String indicates the corresponding table column is a string
+	String
+)
+
+// FieldMeta contains the meta information for decoding and applying feature column on
+// a particular field. A field is a selected column of a SQL result.
+//
+// DType indicates the data type for a field. For example: Int, Float, String.
+//
+// Delimiter indicates the decoding method of a field. For example, the field may
+// contain a string like "1,23,42" which represent a 3-D tensor [1, 23, 42].
+//
+// Shape indicates the shape of the tensor represented for a field. For exmaple, the
+// field may contain a string like "1,23,42" which represent a 3-D tensor, the shape
+// will be [3].
+//
+// IsSparse indicates the type of tensor for a field. True means the tensor is a sparse tensor.
+//
+// FeatureColumn indicates the feature column to be applied on the field. Please refer to
+// github.com/sql-machine-learning/sqlflow/sql/columns for detailed explanation of feature columns.
+// Please be aware that a field can be used by multiple feature columns. For example, the "zipcode"
+// field can be used by CROSS(.) feature and EMBEDDING(CATEGORYID(.)).
+type FieldMeta struct {
+	DType         FieldType               `json:"dtype"`     // e.g. "float", "int32"
+	Delimiter     string                  `json:"delimiter"` // e.g. ","
+	Shape         []int                   `json:"shape"`     // e.g. [1], [1 2 3]
+	IsSparse      bool                    `json:"is_sparse"` // e.g. false
+	FeatureColumn []columns.FeatureColumn `json:"-"`         // e.g. [CrossColumn, EmbeddingColumn(CategoryIDColumn)]
```

Comment:
suggestion: define IR struct like:

```go
type FieldMeta struct {
    FieldNasme string,
    DType FieldType,
    Delimiter string,
    Shape []int,
    IsSparse bool,
}

type FeatureColumn struct {
    FieldMetas []*FieldMeta
    Keys []string,  // CATEGORY_ID(c1): key is c1, EMBEDDING(CATEGORY_ID(c2)) key is c1, CROSS(EMBEDDING(c1), c2): key is c1, c2
}

type TrainIR struct {
...
Feature map[string]map[string]FeatureColumn
}
```

this will much like the struct of column clause, which DENSE/SPARSE defines the `FieldMeta`, furthermore, the `FieldMeta` is derivated when feature derivation is enabled.

```sql
COLUMN EMBEDDING(CATEGORY_ID(SPARSE(c1, 1000, COMMA), 1000), 128, sum),
NUMERIC(DENSE(c2, 128, COMMA), [128])
```